lib/java/os-environment/src/main/java/org/enso/os/environment/chdir/WorkingDirectory.java

```diff
@@ -20,9 +25,70 @@ public static WorkingDirectory getInstance() {
     return INSTANCE;
   }
 
+  /**
+   * In order for the wdir change to be reflected by the JVM, it must be called before {@link
+   * java.io.File} class is loaded.
+   *
+   * <p>In general, try to avoid accessing anything from {@code java.io} before this method is
+   * called.
+   *
+   * @param path Absolute or relative path.
+   * @return True if the working directory was changed, false if an error occurred.
+   */
   public abstract boolean changeWorkingDir(String path);
 
   public abstract String currentWorkingDir();
 
+  /**
+   * Returns true if the given {@code file} exists in the given {@code dir}.
+   *
+   * @param dir Path to the directory. Absolute or relative. Not null. Does not have to end with
+   *     path separator.
+   * @param file Name of the file in the directory. Not null.
+   * @return true if the file exists in the directory, false otherwise.
+   */
   public abstract boolean exists(String dir, String file);
+
+  /**
+   * Attempts to find project root directory.
+   *
+   * @param path Can be absolute or relative. Not null.
+   * @return null if project root was not found, a canonical path otherwise.
+   */
+  public String findProjectRoot(String path) {
```

Comment:
Mixing `abstract` and implementation methods in a single class.  Do you want `findProjectRoot` to be ever overriden? 

```suggestion
  public final String findProjectRoot(String path) {
```

If not, make it final.

Reply:
Yes, this is an oversight, applying your suggestion...